File: marmoraria-app/src/services/firebase_service.py
import requests
import json
import os
import datetime
import socket

# =========================================================
# ==================== CONFIGURAÇÕES ======================
# =========================================================

PROJECT_ID = "marmoraria-app"
BASE_URL = f"https://firestore.googleapis.com/v1/projects/{PROJECT_ID}/databases/(default)/documents"
TIMEOUT = 6  # timeout seguro para requests

# =========================================================
# ===================== CONVERSORES =======================
# =========================================================

def _converter_para_firestore(dados: dict):
    fields = {}

    for k, v in dados.items():
        if v is None:
            fields[k] = {"nullValue": None}

        elif isinstance(v, bool):
            fields[k] = {"booleanValue": v}

        elif isinstance(v, int):
            fields[k] = {"integerValue": str(v)}

        elif isinstance(v, float):
            fields[k] = {"doubleValue": v}

        elif isinstance(v, list):
            values = []
            for item in v:
                if isinstance(item, dict):
                    values.append({
                        "mapValue": {
                            "fields": _converter_para_firestore(item)["fields"]
                        }
                    })
                else:
                    values.append({"stringValue": str(item)})

            fields[k] = {"arrayValue": {"values": values}}

        elif isinstance(v, dict):
            fields[k] = {
                "mapValue": {
                    "fields": _converter_para_firestore(v)["fields"]
                }
            }

        else:
            fields[k] = {"stringValue": str(v)}

    return {"fields": fields}


def _extrair_valor(v):
    if "stringValue" in v:
        return v["stringValue"]
    if "integerValue" in v:
        return int(v["integerValue"])
    if "doubleValue" in v:
        return float(v["doubleValue"])
    if "booleanValue" in v:
        return v["booleanValue"]
    if "nullValue" in v:
        return None
    if "arrayValue" in v:
        return [_extrair_valor(x) for x in v["arrayValue"].get("values", [])]
    if "mapValue" in v:
        return {
            k: _extrair_valor(v2)
            for k, v2 in v["mapValue"].get("fields", {}).items()
        }
    return None


def _converter_de_firestore(doc: dict):
    name = doc.get("name", "")
    obj = {"id": name.split("/")[-1]}

    for k, v in doc.get("fields", {}).items():
        obj[k] = _extrair_valor(v)

    return obj

# =========================================================
# ======================= CONEXÃO =========================
# =========================================================

def verificar_conexao():
    try:
        socket.create_connection(("8.8.8.8", 53), timeout=2)
        return True
    except Exception:
        return False

# =========================================================
# ========================= CRUD ==========================
# =========================================================

def get_document(collection, doc_id):
    if not verificar_conexao():
        return None
    try:
        res = requests.get(
            f"{BASE_URL}/{collection}/{doc_id}",
            timeout=TIMEOUT
        )
        if res.status_code == 200:
            return _converter_de_firestore(res.json())
    except Exception as e:
        print(f"Erro ao buscar documento {collection}/{doc_id}: {e}")
    return None


def get_collection(collection):
    if not verificar_conexao():
        return []
    try:
        res = requests.get(
            f"{BASE_URL}/{collection}?pageSize=300",
            timeout=TIMEOUT
        )
        if res.status_code == 200 and "documents" in res.json():
            return [_converter_de_firestore(d) for d in res.json()["documents"]]
    except Exception:
        pass
    return []


def add_document(collection, dados):
    try:
        res = requests.post(
            f"{BASE_URL}/{collection}",
            json=_converter_para_firestore(dados),
            timeout=TIMEOUT
        )
        return res.status_code == 200
    except Exception:
        return False


def update_document(collection, doc_id, dados):
    try:
        corpo = _converter_para_firestore(dados)
        mask = "&".join([f"updateMask.fieldPaths={k}" for k in dados.keys()])
        url = f"{BASE_URL}/{collection}/{doc_id}?{mask}"

        res = requests.patch(url, json=corpo, timeout=TIMEOUT)
        return res.status_code == 200
    except Exception:
        return False


def delete_document(collection, doc_id):
    try:
        res = requests.delete(
            f"{BASE_URL}/{collection}/{doc_id}",
            timeout=TIMEOUT
        )
        return res.status_code == 200
    except Exception:
        return False

# =========================================================
# ====================== FINANCEIRO =======================
# =========================================================

def get_saldo_caixa():
    saldo = 0.0
    for m in get_collection("financeiro"):
        try:
            v = float(str(m.get("valor", 0)).replace(",", "."))
            tipo = str(m.get("tipo", "")).upper()
            if tipo in ("ENTRADA", "RECEITA"):
                saldo += v
            elif tipo in ("SAIDA", "DESPESA"):
                saldo -= v
        except Exception:
            pass
    return saldo


def get_dividas_pendentes():
    return [
        d for d in get_collection("financeiro")
        if str(d.get("tipo", "")).upper() in ("SAIDA", "DESPESA")
        and d.get("status") != "Pago"
    ]


def get_extrato_lista():
    return sorted(
        get_collection("financeiro"),
        key=lambda x: x.get("data", ""),
        reverse=True
    )


def add_divida_fixa(dados):
    dados["tipo"] = "Saida"
    dados.setdefault("status", "Pendente")
    return add_document("financeiro", dados)


def pagar_divida_fixa(item):
    item["status"] = "Pago"
    item["data_pagamento"] = datetime.datetime.now().isoformat()
    return update_document("financeiro", item["id"], item)


def delete_divida_fixa(id_doc):
    return delete_document("financeiro", id_doc)


def update_divida_fixa(id_doc, dados):
    return update_document("financeiro", id_doc, dados)


def add_movimentacao(tipo, valor, descricao, origem="Manual"):
    dados = {
        "tipo": tipo,
        "valor": float(str(valor).replace(",", ".")),
        "descricao": descricao,
        "origem": origem,
        "data": datetime.datetime.now().isoformat(),
        "status": "Pago" if tipo == "Saida" else "Recebido",
    }
    return add_document("financeiro", dados)


def update_movimentacao(id_doc, dados):
    return update_document("financeiro", id_doc, dados)


def delete_movimentacao(id_doc):
    return delete_document("financeiro", id_doc)

# =========================================================
# ===================== ORÇAMENTOS ========================
# =========================================================

def get_orcamentos_lista():
    return get_collection("orcamentos")


def get_orcamentos_finalizados_nao_pagos():
    return [
        o for o in get_collection("orcamentos")
        if o.get("status") == "Finalizado" and not o.get("pago")
    ]


def receber_orcamento(item):
    item["pago"] = True
    item["status_pagamento"] = "Pago"
    item["status"] = "Finalizado"

    update_document("orcamentos", item["id"], item)

    valor_total = float(str(item.get("total_geral", 0)).replace(",", "."))

    return add_movimentacao(
        "Entrada",
        valor_total,
        f"Receb. Orç: {item.get('cliente_nome')}",
        "Vendas",
    )

# =========================================================
# =================== USUÁRIOS / LOGIN ====================
# =========================================================

def get_user_doc_by_email(email):
    for u in get_collection("users"):
        if u.get("email") == email:
            return u
    return None


def verify_user_password(email, password):
    user = get_user_doc_by_email(email)
    if user:
        senha_db = str(user.get("senha") or user.get("password") or "")
        return senha_db == str(password)
    return False


def initialize_firebase():
    return True


def get_collection_count(collection):
    return len(get_collection(collection))


def get_orcamentos_by_status(status):
    return [
        o for o in get_collection("orcamentos")
        if o.get("status") == status
    ]

File: lib/core/services/firebase_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

class FirebaseService {
  static const String projectId = 'marmoraria-app';
  static const String baseUrl =
      'https://firestore.googleapis.com/v1/projects/$projectId/databases/(default)/documents';

  /* ======================================================
   * CONEXÃO
   * ====================================================== */

  static Future<bool> verificarConexao() async {
    try {
      final socket = await Socket.connect('8.8.8.8', 53,
          timeout: const Duration(seconds: 2));
      socket.destroy();
      return true;
    } catch (_) {
      return false;
    }
  }

  /* ======================================================
   * CONVERSORES
   * ====================================================== */

  static Map<String, dynamic> _toFirestore(Map<String, dynamic> data) {
    final Map<String, dynamic> fields = {};

    data.forEach((key, value) {
      if (value == null) {
        fields[key] = {'nullValue': null};
      } else if (value is bool) {
        fields[key] = {'booleanValue': value};
      } else if (value is int) {
        fields[key] = {'integerValue': value.toString()};
      } else if (value is double) {
        fields[key] = {'doubleValue': value};
      } else if (value is List) {
        fields[key] = {
          'arrayValue': {
            'values': value.map((v) {
              if (v is Map<String, dynamic>) {
                return {'mapValue': _toFirestore(v)};
              }
              return {'stringValue': v.toString()};
            }).toList()
          }
        };
      } else if (value is Map<String, dynamic>) {
        fields[key] = {'mapValue': _toFirestore(value)};
      } else {
        fields[key] = {'stringValue': value.toString()};
      }
    });

    return {'fields': fields};
  }

  static dynamic _fromValue(Map<String, dynamic> v) {
    if (v.containsKey('stringValue')) return v['stringValue'];
    if (v.containsKey('integerValue')) return int.parse(v['integerValue']);
    if (v.containsKey('doubleValue')) return v['doubleValue'];
    if (v.containsKey('booleanValue')) return v['booleanValue'];
    if (v.containsKey('nullValue')) return null;

    if (v.containsKey('arrayValue')) {
      return (v['arrayValue']['values'] ?? [])
          .map((e) => _fromValue(e))
          .toList();
    }

    if (v.containsKey('mapValue')) {
      final Map<String, dynamic> map = {};
      (v['mapValue']['fields'] ?? {}).forEach((k, val) {
        map[k] = _fromValue(val);
      });
      return map;
    }

    return null;
  }

  static Map<String, dynamic> _fromFirestore(Map<String, dynamic> doc) {
    final Map<String, dynamic> obj = {};
    obj['id'] = doc['name'].split('/').last;

    (doc['fields'] ?? {}).forEach((k, v) {
      obj[k] = _fromValue(v);
    });

    return obj;
  }

  /* ======================================================
   * CRUD BASE
   * ====================================================== */

  static Future<List<Map<String, dynamic>>> getCollection(
      String collection) async {
    if (!await verificarConexao()) return [];

    final res =
        await http.get(Uri.parse('$baseUrl/$collection?pageSize=300'));

    if (res.statusCode != 200) return [];

    final body = jsonDecode(res.body);
    if (!body.containsKey('documents')) return [];

    return body['documents']
        .map<Map<String, dynamic>>((d) => _fromFirestore(d))
        .toList();
  }

  static Future<bool> addDocument(
      String collection, Map<String, dynamic> data) async {
    final res = await http.post(
      Uri.parse('$baseUrl/$collection'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(_toFirestore(data)),
    );
    return res.statusCode == 200;
  }

  static Future<bool> updateDocument(
      String collection, String id, Map<String, dynamic> data) async {
    final mask = data.keys.map((k) => 'updateMask.fieldPaths=$k').join('&');
    final url = '$baseUrl/$collection/$id?$mask';

    final res = await http.patch(
      Uri.parse(url),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(_toFirestore(data)),
    );
    return res.statusCode == 200;
  }

  static Future<bool> deleteDocument(
      String collection, String id) async {
    final res =
        await http.delete(Uri.parse('$baseUrl/$collection/$id'));
    return res.statusCode == 200;
  }

  /* ======================================================
   * FINANCEIRO
   * ====================================================== */

  static Future<double> getSaldoCaixa() async {
    final movs = await getCollection('financeiro');
    double saldo = 0;

    for (final m in movs) {
      final valor = double.tryParse(
              m['valor'].toString().replaceAll(',', '.')) ??
          0;
      final tipo = (m['tipo'] ?? '').toString().toUpperCase();

      if (['ENTRADA', 'RECEITA'].contains(tipo)) {
        saldo += valor;
      } else if (['SAIDA', 'DESPESA'].contains(tipo)) {
        saldo -= valor;
      }
    }
    return saldo;
  }

  static Future<List<Map<String, dynamic>>> getExtrato() async {
    final dados = await getCollection('financeiro');
    dados.sort((a, b) =>
        (b['data'] ?? '').compareTo(a['data'] ?? ''));
    return dados;
  }

  static Future<bool> addMovimentacao(
      String tipo, double valor, String descricao,
      {String origem = 'Manual'}) {
    return addDocument('financeiro', {
      'tipo': tipo,
      'valor': valor,
      'descricao': descricao,
      'origem': origem,
      'data': DateTime.now().toIso8601String(),
      'status': tipo == 'Saida' ? 'Pago' : 'Recebido',
    });
  }

  /* ======================================================
   * PRODUÇÃO
   * ====================================================== */

  static Future<List<Map<String, dynamic>>> getOrcamentosByStatus(
      String status) async {
    final todos = await getCollection('orcamentos');
    return todos.where((o) => o['status'] == status).toList();
  }
}
